### PR TITLE
Issue 27 - Fix off by one error 

### PR DESCRIPTION
**Pull request by mantono**

### CODE REVIEW

File: src/main/kotlin/engine/kafka_one/JobStatuses.kt [hunk 1]
```diff
@@ -4,8 +4,15 @@ import franz.JobStatus
 import org.apache.kafka.clients.consumer.ConsumerRecord
 import org.apache.kafka.clients.consumer.OffsetAndMetadata
 import org.apache.kafka.common.TopicPartition
+import java.lang.IllegalStateException
 
-private fun <K, V> Map<K, V>.getOrFail(k: K) = get(k)!!
+private fun <K, V> Map<K, V>.getOrFail(k: K): V {
+    return try {
+        get(k)!!
+    } catch(e: NullPointerException) {
```
Comment (williamhogman) [inline src/main/kotlin/engine/kafka_one/JobStatuses.kt]: Write this for as an If null statement instead.

File: src/main/kotlin/engine/kafka_one/JobStatuses.kt [hunk 2]
```diff
@@ -29,11 +36,11 @@ data class JobStatuses<T, U>(
         copy(
                 jobStatuses = jobStatuses.filterKeys { (topicPartition, offset) ->
                     val committedOffset = committed[topicPartition]?.offset() ?: -1
-                    offset > committedOffset
+                    offset >= committedOffset
```
Comment (williamhogman) [inline src/main/kotlin/engine/kafka_one/JobStatuses.kt:34]: This

File: src/main/kotlin/engine/kafka_one/JobStatuses.kt [hunk 3]
```diff
@@ -29,11 +36,11 @@ data class JobStatuses<T, U>(
...
                 },
                 records = records.filterValues {
                     val committedOffset = committed[it.topicPartition()]?.offset() ?: -1
-                    it.offset() > committedOffset
+                    it.offset() >= committedOffset
```
Comment (williamhogman) [inline src/main/kotlin/engine/kafka_one/JobStatuses.kt:38]: makes

File: src/main/kotlin/engine/kafka_one/JobStatuses.kt [hunk 4]
```diff
@@ -42,7 +49,10 @@ data class JobStatuses<T, U>(
...
             changeBatch(jobs.map { it.jobId() }, JobStatus.Incomplete)
                     .copy(records = records + jobs.map { it.jobId() to it })
-    fun rescheduleTransientFailures() = jobStatuses.filterValues { it.mayRetry() }.keys.let {
-        changeBatch(it, JobStatus.Retry) to it.map(records::getOrFail)
-    }
+    fun rescheduleTransientFailures(): Pair<JobStatuses<T, U>, List<ConsumerRecord<T, U>>> =
+        jobStatuses
+            .filterValues { it.mayRetry() }
+            .keys.let { jobIds: Set<JobId> ->
+            changeBatch(jobIds, JobStatus.Retry) to jobIds.map(records::getOrFail)
```
Comment (williamhogman) [inline src/main/kotlin/engine/kafka_one/JobStatuses.kt:51]: sense

**Review by williamhogman — CHANGES_REQUESTED** [review]
avoid throwing and catching an exception in the same function. Otherwise ok to merge

File: src/main/kotlin/engine/kafka_one/JobStatuses.kt (hunk 1 above)
Comment (mantono) [reply src/main/kotlin/engine/kafka_one/JobStatuses.kt]: Yeah, right. That will be prettier and make more sense :ok_hand:

**Pull request merged**